// react-shaders/dist/index.mjs
var __defProp = Object.defineProperty;
var __getOwnPropSymbols = Object.getOwnPropertySymbols;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __propIsEnum = Object.prototype.propertyIsEnumerable;
var __pow = Math.pow;
var __defNormalProp = (obj, key, value) => key in obj ? __defProp(obj, key, { enumerable: true, configurable: true, writable: true, value }) : obj[key] = value;
var __spreadValues = (a, b) => {
  for (var prop in b || (b = {}))
    if (__hasOwnProp.call(b, prop))
      __defNormalProp(a, prop, b[prop]);
  if (__getOwnPropSymbols)
    for (var prop of __getOwnPropSymbols(b)) {
      if (__propIsEnum.call(b, prop))
        __defNormalProp(a, prop, b[prop]);
    }
  return a;
};
var __async = (__this, __arguments, generator) => {
  return new Promise((resolve, reject) => {
    var fulfilled = (value) => {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    };
    var rejected = (value) => {
      try {
        step(generator.throw(value));
      } catch (e) {
        reject(e);
      }
    };
    var step = (x) => x.done ? resolve(x.value) : Promise.resolve(x.value).then(fulfilled, rejected);
    step((generator = generator.apply(__this, __arguments)).next());
  });
};

// core/index.tsx
import React, { Component } from "react";

// core/logging.ts
var RSLOG = (text) => `react-shaders: ${text}`;

// core/texture.ts
var NearestFilter = 9728;
var LinearFilter = 9729;
var NearestMipMapNearestFilter = 9984;
var LinearMipMapNearestFilter = 9985;
var NearestMipMapLinearFilter = 9986;
var LinearMipMapLinearFilter = 9987;
var ClampToEdgeWrapping = 33071;
var MirroredRepeatWrapping = 33648;
var RepeatWrapping = 10497;
var isPowerOf2 = (value) => (value & value - 1) == 0;
var floorPowerOfTwo = (value) => __pow(2, Math.floor(Math.log(value) / Math.LN2));
var textureNeedsGenerateMipmaps = (texture, isPowerOfTwo) => isPowerOfTwo && texture.minFilter !== NearestFilter && texture.minFilter !== LinearFilter;
var textureNeedsPowerOfTwo = (texture) => {
  if (texture.wrapS !== ClampToEdgeWrapping || texture.wrapT !== ClampToEdgeWrapping)
    return true;
  if (texture.minFilter !== NearestFilter && texture.minFilter !== LinearFilter)
    return true;
  return false;
};
var Texture = class {
  constructor(gl) {
    this.isLoaded = false;
    this.isVideo = false;
    this.flipY = -1;
    this.width = 0;
    this.height = 0;
    this._webglTexture = null;
    this.updateTexture = (texture, video, flipY) => {
      const { gl } = this;
      const level = 0;
      const internalFormat = gl.RGBA;
      const srcFormat = gl.RGBA;
      const srcType = gl.UNSIGNED_BYTE;
      gl.bindTexture(gl.TEXTURE_2D, texture);
      gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, flipY);
      gl.texImage2D(
        gl.TEXTURE_2D,
        level,
        internalFormat,
        srcFormat,
        srcType,
        video
      );
    };
    this.setupVideo = (url) => {
      const video = document.createElement("video");
      let playing = false;
      let timeupdate = false;
      video.autoplay = true;
      video.muted = true;
      video.loop = true;
      video.crossOrigin = "anonymous";
      const checkReady = () => {
        if (playing && timeupdate) {
          this.isLoaded = true;
        }
      };
      video.addEventListener(
        "playing",
        () => {
          playing = true;
          this.width = video.videoWidth || 0;
          this.height = video.videoHeight || 0;
          checkReady();
        },
        true
      );
      video.addEventListener(
        "timeupdate",
        () => {
          timeupdate = true;
          checkReady();
        },
        true
      );
      video.src = url;
      return video;
    };
    this.makePowerOfTwo = (image) => {
      if (image instanceof HTMLImageElement || image instanceof HTMLCanvasElement || image instanceof ImageBitmap) {
        if (this.pow2canvas === void 0)
          this.pow2canvas = document.createElement("canvas");
        this.pow2canvas.width = floorPowerOfTwo(image.width);
        this.pow2canvas.height = floorPowerOfTwo(image.height);
        const context = this.pow2canvas.getContext("2d");
        context == null ? void 0 : context.drawImage(
          image,
          0,
          0,
          this.pow2canvas.width,
          this.pow2canvas.height
        );
        console.warn(
          RSLOG(
            `Image is not power of two ${image.width} x ${image.height}. Resized to ${this.pow2canvas.width} x ${this.pow2canvas.height};`
          )
        );
        return this.pow2canvas;
      }
      return image;
    };
    this.load = (textureArgs) => __async(this, null, function* () {
      const { gl } = this;
      const {
        url,
        wrapS,
        wrapT,
        minFilter,
        magFilter,
        flipY = -1
      } = textureArgs;
      if (!url) {
        return Promise.reject(
          new Error(
            RSLOG(
              "Missing url, please make sure to pass the url of your texture { url: ... }"
            )
          )
        );
      }
      const isImage = /(\.jpg|\.jpeg|\.png|\.gif|\.bmp)$/i.exec(url);
      const isVideo = /(\.mp4|\.3gp|\.webm|\.ogv)$/i.exec(url);
      if (isImage === null && isVideo === null) {
        return Promise.reject(
          new Error(
            RSLOG(
              `Please upload a video or an image with a valid format (url: ${url})`
            )
          )
        );
      }
      Object.assign(this, {
        url,
        wrapS,
        wrapT,
        minFilter,
        magFilter,
        flipY
      });
      const level = 0;
      const internalFormat = gl.RGBA;
      const width = 1;
      const height = 1;
      const border = 0;
      const srcFormat = gl.RGBA;
      const srcType = gl.UNSIGNED_BYTE;
      const pixel = new Uint8Array([255, 255, 255, 0]);
      const texture = gl.createTexture();
      gl.bindTexture(gl.TEXTURE_2D, texture);
      gl.texImage2D(
        gl.TEXTURE_2D,
        level,
        internalFormat,
        width,
        height,
        border,
        srcFormat,
        srcType,
        pixel
      );
      if (isVideo) {
        const video = this.setupVideo(url);
        gl.texParameteri(
          gl.TEXTURE_2D,
          gl.TEXTURE_WRAP_S,
          gl.CLAMP_TO_EDGE
        );
        gl.texParameteri(
          gl.TEXTURE_2D,
          gl.TEXTURE_WRAP_T,
          gl.CLAMP_TO_EDGE
        );
        gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.LINEAR);
        this._webglTexture = texture;
        this.source = video;
        this.isVideo = true;
        return video.play().then(() => this);
      }
      function loadImage() {
        return __async(this, null, function* () {
          return new Promise((resolve, reject) => {
            const image2 = new Image();
            image2.crossOrigin = "anonymous";
            image2.onload = () => {
              resolve(image2);
            };
            image2.onerror = () => {
              reject(new Error(RSLOG(`failed loading url: ${url}`)));
            };
            image2.src = url != null ? url : "";
          });
        });
      }
      let image = yield loadImage();
      let isPowerOfTwoImage = isPowerOf2(image.width) && isPowerOf2(image.height);
      if (textureNeedsPowerOfTwo(textureArgs) && !isPowerOfTwoImage) {
        image = this.makePowerOfTwo(image);
        isPowerOfTwoImage = true;
      }
      gl.bindTexture(gl.TEXTURE_2D, texture);
      gl.pixelStorei(gl.UNPACK_FLIP_Y_WEBGL, flipY);
      gl.texImage2D(
        gl.TEXTURE_2D,
        level,
        internalFormat,
        srcFormat,
        srcType,
        image
      );
      if (textureNeedsGenerateMipmaps(textureArgs, isPowerOfTwoImage)) {
        gl.generateMipmap(gl.TEXTURE_2D);
      }
      gl.texParameteri(
        gl.TEXTURE_2D,
        gl.TEXTURE_WRAP_S,
        this.wrapS || RepeatWrapping
      );
      gl.texParameteri(
        gl.TEXTURE_2D,
        gl.TEXTURE_WRAP_T,
        this.wrapT || RepeatWrapping
      );
      gl.texParameteri(
        gl.TEXTURE_2D,
        gl.TEXTURE_MIN_FILTER,
        this.minFilter || LinearMipMapLinearFilter
      );
      gl.texParameteri(
        gl.TEXTURE_2D,
        gl.TEXTURE_MAG_FILTER,
        this.magFilter || LinearFilter
      );
      this._webglTexture = texture;
      this.source = image;
      this.isVideo = false;
      this.isLoaded = true;
      this.width = image.width || 0;
      this.height = image.height || 0;
      return this;
    });
    this.gl = gl;
  }
};

// core/uniforms.ts
var processUniform = (gl, location, t, value) => {
  switch (t) {
    case "1f":
      gl.uniform1f(location, value);
      break;
    case "2f":
      gl.uniform2f(location, value[0], value[1]);
      break;
    case "3f":
      gl.uniform3f(location, value[0], value[1], value[2]);
      break;
    case "4f":
      gl.uniform4f(location, value[0], value[1], value[2], value[3]);
      break;
    case "1i":
      gl.uniform1i(location, value);
      break;
    case "2i":
      gl.uniform2i(location, value[0], value[1]);
      break;
    case "3i":
      gl.uniform3i(location, value[0], value[1], value[2]);
      break;
    case "4i":
      gl.uniform3i(location, value[0], value[1], value[2], value[3]);
      break;
    case "1iv":
      gl.uniform1iv(location, value);
      break;
    case "2iv":
      gl.uniform2iv(location, value);
      break;
    case "3iv":
      gl.uniform3iv(location, value);
      break;
    case "4iv":
      gl.uniform4iv(location, value);
      break;
    case "1fv":
      gl.uniform1fv(location, value);
      break;
    case "2fv":
      gl.uniform2fv(location, value);
      break;
    case "3fv":
      gl.uniform3fv(location, value);
      break;
    case "4fv":
      gl.uniform4fv(location, value);
      break;
    case "Matrix2fv":
      gl.uniformMatrix2fv(location, false, value);
      break;
    case "Matrix3fv":
      gl.uniformMatrix3fv(location, false, value);
      break;
    case "Matrix4fv":
      gl.uniformMatrix4fv(location, false, value);
      break;
    default:
      break;
  }
};
var uniformTypeToGLSLType = (t) => {
  switch (t) {
    case "1f":
      return "float";
    case "2f":
      return "vec2";
    case "3f":
      return "vec3";
    case "4f":
      return "vec4";
    case "1i":
      return "int";
    case "2i":
      return "ivec2";
    case "3i":
      return "ivec3";
    case "4i":
      return "ivec4";
    case "1iv":
      return "int";
    case "2iv":
      return "ivec2";
    case "3iv":
      return "ivec3";
    case "4iv":
      return "ivec4";
    case "1fv":
      return "float";
    case "2fv":
      return "vec2";
    case "3fv":
      return "vec3";
    case "4fv":
      return "vec4";
    case "Matrix2fv":
      return "mat2";
    case "Matrix3fv":
      return "mat3";
    case "Matrix4fv":
      return "mat4";
    default:
      console.error(
        RSLOG(
          `The uniform type "${t}" is not valid, please make sure your uniform type is valid`
        )
      );
  }
};

// core/index.tsx
var PRECISIONS = ["lowp", "mediump", "highp"];
var FS_MAIN_SHADER = `
void main(void){
    vec4 color = vec4(0.0,0.0,0.0,1.0);
    mainImage( color, gl_FragCoord.xy );
    gl_FragColor = color;
}`;
var BASIC_FS = `void mainImage( out vec4 fragColor, in vec2 fragCoord ) {
    vec2 uv = fragCoord/iResolution.xy;
    vec3 col = 0.5 + 0.5*cos(iTime+uv.xyx+vec3(0,2,4));
    fragColor = vec4(col,1.0);
}`;
var BASIC_VS = `attribute vec3 aVertexPosition;
void main(void) {
    gl_Position = vec4(aVertexPosition, 1.0);
}`;
var UNIFORM_TIME = "iTime";
var UNIFORM_TIMEDELTA = "iTimeDelta";
var UNIFORM_DATE = "iDate";
var UNIFORM_FRAME = "iFrame";
var UNIFORM_MOUSE = "iMouse";
var UNIFORM_RESOLUTION = "iResolution";
var UNIFORM_CHANNEL = "iChannel";
var UNIFORM_CHANNELRESOLUTION = "iChannelResolution";
var UNIFORM_DEVICEORIENTATION = "iDeviceOrientation";
var latestPointerClientCoords = (e) => {
  return [
    // @ts-expect-error TODO: Deal with this.
    e.clientX || e.changedTouches[0].clientX,
    // @ts-expect-error TODO: Deal with this.
    clientY = e.clientY || e.changedTouches[0].clientY
  ];
};
var lerpVal = (v0, v1, t) => v0 * (1 - t) + v1 * t;
var insertStringAtIndex = (currentString, string, index) => index > 0 ? currentString.substring(0, index) + string + currentString.substring(index, currentString.length) : string + currentString;
var Shader = class extends Component {
  constructor(props) {
    super(props);
    this.componentDidMount = () => {
      this.initWebGL();
      const { fs, vs, clearColor = [0, 0, 0, 1] } = this.props;
      const { gl } = this;
      if (gl && this.canvas) {
        gl.clearColor(...clearColor);
        gl.clearDepth(1);
        gl.enable(gl.DEPTH_TEST);
        gl.depthFunc(gl.LEQUAL);
        gl.viewport(0, 0, this.canvas.width, this.canvas.height);
        this.canvas.height = this.canvas.clientHeight;
        this.canvas.width = this.canvas.clientWidth;
        this.processCustomUniforms();
        this.processTextures();
        const shaders = this.preProcessShaders(
          fs || BASIC_FS,
          vs || BASIC_VS
        );
        this.initShaders(shaders);
        this.initBuffers();
        this.drawScene();
        this.addEventListeners();
        this.onResize();
      }
    };
    this.shouldComponentUpdate = () => false;
    this.setupChannelRes = ({ width, height }, id) => {
      const { devicePixelRatio = 1 } = this.props;
      this.uniforms.iChannelResolution.value[id * 3] = width * devicePixelRatio;
      this.uniforms.iChannelResolution.value[id * 3 + 1] = height * devicePixelRatio;
      this.uniforms.iChannelResolution.value[id * 3 + 2] = 0;
    };
    this.initWebGL = () => {
      var _a, _b;
      const { contextAttributes } = this.props;
      if (!this.canvas)
        return;
      this.gl = this.canvas.getContext("webgl", contextAttributes) || this.canvas.getContext(
        "experimental-webgl",
        contextAttributes
      );
      (_a = this.gl) == null ? void 0 : _a.getExtension("OES_standard_derivatives");
      (_b = this.gl) == null ? void 0 : _b.getExtension("EXT_shader_texture_lod");
    };
    this.initBuffers = () => {
      var _a;
      const { gl } = this;
      this.squareVerticesBuffer = gl == null ? void 0 : gl.createBuffer();
      gl == null ? void 0 : gl.bindBuffer(gl.ARRAY_BUFFER, (_a = this.squareVerticesBuffer) != null ? _a : null);
      const vertices = [
        1,
        1,
        0,
        -1,
        1,
        0,
        1,
        -1,
        0,
        -1,
        -1,
        0
      ];
      gl == null ? void 0 : gl.bufferData(
        gl.ARRAY_BUFFER,
        new Float32Array(vertices),
        gl.STATIC_DRAW
      );
    };
    this.addEventListeners = () => {
      var _a, _b;
      const options = {
        passive: true
      };
      if (((_a = this.uniforms.iMouse) == null ? void 0 : _a.isNeeded) && this.canvas) {
        this.canvas.addEventListener("mousemove", this.mouseMove, options);
        this.canvas.addEventListener("mouseout", this.mouseUp, options);
        this.canvas.addEventListener("mouseup", this.mouseUp, options);
        this.canvas.addEventListener("mousedown", this.mouseDown, options);
        this.canvas.addEventListener("touchmove", this.mouseMove, options);
        this.canvas.addEventListener("touchend", this.mouseUp, options);
        this.canvas.addEventListener(
          "touchstart",
          this.mouseDown,
          options
        );
      }
      if ((_b = this.uniforms.iDeviceOrientation) == null ? void 0 : _b.isNeeded) {
        window.addEventListener(
          "deviceorientation",
          this.onDeviceOrientationChange,
          options
        );
      }
      window.addEventListener("resize", this.onResize, options);
    };
    this.removeEventListeners = () => {
      var _a, _b;
      const options = {
        passive: true
      };
      if (((_a = this.uniforms.iMouse) == null ? void 0 : _a.isNeeded) && this.canvas) {
        this.canvas.removeEventListener(
          "mousemove",
          this.mouseMove,
          options
        );
        this.canvas.removeEventListener("mouseout", this.mouseUp, options);
        this.canvas.removeEventListener("mouseup", this.mouseUp, options);
        this.canvas.removeEventListener(
          "mousedown",
          this.mouseDown,
          options
        );
        this.canvas.removeEventListener(
          "touchmove",
          this.mouseMove,
          options
        );
        this.canvas.removeEventListener("touchend", this.mouseUp, options);
        this.canvas.removeEventListener(
          "touchstart",
          this.mouseDown,
          options
        );
      }
      if ((_b = this.uniforms.iDeviceOrientation) == null ? void 0 : _b.isNeeded) {
        window.removeEventListener(
          "deviceorientation",
          this.onDeviceOrientationChange,
          options
        );
      }
      window.removeEventListener("resize", this.onResize, options);
    };
    this.onDeviceOrientationChange = ({
      alpha,
      beta,
      gamma
    }) => {
      this.uniforms.iDeviceOrientation.value = [
        alpha != null ? alpha : 0,
        beta != null ? beta : 0,
        gamma != null ? gamma : 0,
        window.orientation || 0
      ];
    };
    this.mouseDown = (e) => {
      var _a, _b, _c, _d, _e, _f;
      const [clientX, clientY2] = latestPointerClientCoords(e);
      const mouseX = clientX - ((_b = (_a = this.canvasPosition) == null ? void 0 : _a.left) != null ? _b : 0) - window.pageXOffset;
      const mouseY = ((_d = (_c = this.canvasPosition) == null ? void 0 : _c.height) != null ? _d : 0) - clientY2 - ((_f = (_e = this.canvasPosition) == null ? void 0 : _e.top) != null ? _f : 0) - window.pageYOffset;
      this.mousedown = true;
      this.uniforms.iMouse.value[2] = mouseX;
      this.uniforms.iMouse.value[3] = mouseY;
      this.lastMouseArr[0] = mouseX;
      this.lastMouseArr[1] = mouseY;
    };
    this.mouseMove = (e) => {
      var _a, _b, _c, _d, _e, _f, _g;
      this.canvasPosition = (_a = this.canvas) == null ? void 0 : _a.getBoundingClientRect();
      const { lerp = 1 } = this.props;
      const [clientX, clientY2] = latestPointerClientCoords(e);
      const mouseX = clientX - ((_c = (_b = this.canvasPosition) == null ? void 0 : _b.left) != null ? _c : 0);
      const mouseY = ((_e = (_d = this.canvasPosition) == null ? void 0 : _d.height) != null ? _e : 0) - clientY2 - ((_g = (_f = this.canvasPosition) == null ? void 0 : _f.top) != null ? _g : 0);
      if (lerp !== 1) {
        this.lastMouseArr[0] = mouseX;
        this.lastMouseArr[1] = mouseY;
      } else {
        this.uniforms.iMouse.value[0] = mouseX;
        this.uniforms.iMouse.value[1] = mouseY;
      }
    };
    this.mouseUp = () => {
      this.uniforms.iMouse.value[2] = 0;
      this.uniforms.iMouse.value[3] = 0;
    };
    this.onResize = () => {
      var _a, _b, _c, _d, _e, _f;
      const { gl } = this;
      const { devicePixelRatio = 1 } = this.props;
      if (!gl)
        return;
      this.canvasPosition = (_a = this.canvas) == null ? void 0 : _a.getBoundingClientRect();
      const realToCSSPixels = devicePixelRatio;
      const displayWidth = Math.floor(
        ((_c = (_b = this.canvasPosition) == null ? void 0 : _b.width) != null ? _c : 1) * realToCSSPixels
      );
      const displayHeight = Math.floor(
        ((_e = (_d = this.canvasPosition) == null ? void 0 : _d.height) != null ? _e : 1) * realToCSSPixels
      );
      gl.canvas.width = displayWidth;
      gl.canvas.height = displayHeight;
      if (((_f = this.uniforms.iResolution) == null ? void 0 : _f.isNeeded) && this.shaderProgram) {
        const rUniform = gl.getUniformLocation(
          this.shaderProgram,
          UNIFORM_RESOLUTION
        );
        gl.uniform2fv(rUniform, [gl.canvas.width, gl.canvas.height]);
      }
    };
    this.drawScene = (timestamp) => {
      var _a, _b, _c;
      const { gl } = this;
      const { lerp = 1 } = this.props;
      if (!gl)
        return;
      gl.viewport(0, 0, gl.drawingBufferWidth, gl.drawingBufferHeight);
      gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);
      gl.bindBuffer(gl.ARRAY_BUFFER, (_a = this.squareVerticesBuffer) != null ? _a : null);
      gl.vertexAttribPointer(
        (_b = this.vertexPositionAttribute) != null ? _b : 0,
        3,
        gl.FLOAT,
        false,
        0,
        0
      );
      this.setUniforms(timestamp);
      gl.drawArrays(gl.TRIANGLE_STRIP, 0, 4);
      if (((_c = this.uniforms.iMouse) == null ? void 0 : _c.isNeeded) && lerp !== 1) {
        this.uniforms.iMouse.value[0] = lerpVal(
          // @ts-expect-error TODO: Deal with this.
          this.uniforms.iMouse.value[0],
          // @ts-expect-error TODO: Deal with this.
          this.lastMouseArr[0],
          lerp
        );
        this.uniforms.iMouse.value[1] = lerpVal(
          // @ts-expect-error TODO: Deal with this.
          this.uniforms.iMouse.value[1],
          // @ts-expect-error TODO: Deal with this.
          this.lastMouseArr[1],
          lerp
        );
      }
      this.animFrameId = requestAnimationFrame(this.drawScene);
    };
    this.createShader = (type, shaderCodeAsText) => {
      var _a, _b, _c, _d;
      const { gl } = this;
      if (!gl)
        return null;
      const shader = gl.createShader(type);
      if (!shader)
        return null;
      gl.shaderSource(shader, shaderCodeAsText);
      gl.compileShader(shader);
      if (!gl.getShaderParameter(shader, gl.COMPILE_STATUS)) {
        (_b = (_a = this.props).onWarning) == null ? void 0 : _b.call(
          _a,
          RSLOG(`Error compiling the shader:
${shaderCodeAsText}`)
        );
        const compilationLog = gl.getShaderInfoLog(shader);
        gl.deleteShader(shader);
        (_d = (_c = this.props).onError) == null ? void 0 : _d.call(
          _c,
          RSLOG(`Shader compiler log: ${compilationLog}`)
        );
      }
      return shader;
    };
    this.initShaders = ({ fs, vs }) => {
      var _a, _b;
      const { gl } = this;
      if (!gl)
        return;
      const fragmentShader = this.createShader(gl.FRAGMENT_SHADER, fs);
      const vertexShader = this.createShader(gl.VERTEX_SHADER, vs);
      this.shaderProgram = gl.createProgram();
      if (!this.shaderProgram || !vertexShader || !fragmentShader)
        return;
      gl.attachShader(this.shaderProgram, vertexShader);
      gl.attachShader(this.shaderProgram, fragmentShader);
      gl.linkProgram(this.shaderProgram);
      if (!gl.getProgramParameter(this.shaderProgram, gl.LINK_STATUS)) {
        (_b = (_a = this.props).onError) == null ? void 0 : _b.call(
          _a,
          RSLOG(
            `Unable to initialize the shader program: ${gl.getProgramInfoLog(
              this.shaderProgram
            )}`
          )
        );
        return;
      }
      gl.useProgram(this.shaderProgram);
      this.vertexPositionAttribute = gl.getAttribLocation(
        this.shaderProgram,
        "aVertexPosition"
      );
      gl.enableVertexAttribArray(this.vertexPositionAttribute);
    };
    this.processCustomUniforms = () => {
      const { uniforms } = this.props;
      if (uniforms) {
        Object.keys(uniforms).forEach((name) => {
          var _a;
          const uniform = (_a = this.props.uniforms) == null ? void 0 : _a[name];
          if (!uniform)
            return;
          const { value, type } = uniform;
          const glslType = uniformTypeToGLSLType(type);
          if (!glslType)
            return;
          function isMatrixType(t, v) {
            return t.includes("Matrix") && Array.isArray(v);
          }
          function isVectorListType(t, v) {
            return t.includes("v") && Array.isArray(v) && v.length > parseInt(t.charAt(0));
          }
          const tempObject = {};
          if (isMatrixType(type, value)) {
            const arrayLength = type.length;
            const val = parseInt(type.charAt(arrayLength - 3));
            const numberOfMatrices = Math.floor(
              value.length / (val * val)
            );
            if (value.length > val * val) {
              tempObject.arraySize = `[${numberOfMatrices}]`;
            }
          } else if (isVectorListType(type, value)) {
            tempObject.arraySize = `[${Math.floor(
              value.length / parseInt(type.charAt(0))
            )}]`;
          }
          this.uniforms[name] = __spreadValues({
            type: glslType,
            isNeeded: false,
            value
          }, tempObject);
        });
      }
    };
    this.processTextures = () => {
      const { gl } = this;
      const { textures, onDoneLoadingTextures } = this.props;
      if (!gl)
        return;
      if (textures && textures.length > 0) {
        this.uniforms[`${UNIFORM_CHANNELRESOLUTION}`] = {
          type: "vec3",
          isNeeded: false,
          arraySize: `[${textures.length}]`,
          value: []
        };
        const texturePromisesArr = textures.map(
          (texture, id) => {
            var _a;
            this.uniforms[`${UNIFORM_CHANNEL}${id}`] = {
              type: "sampler2D",
              isNeeded: false
            };
            this.setupChannelRes(texture, id);
            this.texturesArr[id] = new Texture(gl);
            return (_a = this.texturesArr[id]) == null ? void 0 : _a.load(texture, id).then((t) => {
              this.setupChannelRes(t, id);
            });
          }
        );
        Promise.all(texturePromisesArr).then(() => {
          if (onDoneLoadingTextures)
            onDoneLoadingTextures();
        }).catch((e) => {
          var _a, _b;
          (_b = (_a = this.props).onError) == null ? void 0 : _b.call(_a, e);
          if (onDoneLoadingTextures)
            onDoneLoadingTextures();
        });
      } else {
        if (onDoneLoadingTextures)
          onDoneLoadingTextures();
      }
    };
    this.preProcessShaders = (fs, vs) => {
      var _a, _b;
      const { precision, devicePixelRatio = 1 } = this.props;
      const dprString = `#define DPR ${devicePixelRatio.toFixed(1)}
`;
      const isValidPrecision = PRECISIONS.includes(precision != null ? precision : "highp");
      const precisionString = `precision ${isValidPrecision ? precision : PRECISIONS[1]} float;
`;
      if (!isValidPrecision)
        (_b = (_a = this.props).onWarning) == null ? void 0 : _b.call(
          _a,
          RSLOG(
            `wrong precision type ${precision}, please make sure to pass one of a valid precision lowp, mediump, highp, by default you shader precision will be set to highp.`
          )
        );
      let fsString = precisionString.concat(dprString).concat(fs).replace(/texture\(/g, "texture2D(");
      const indexOfPrecisionString = fsString.lastIndexOf(precisionString);
      Object.keys(this.uniforms).forEach((uniform) => {
        if (fs.includes(uniform)) {
          const u = this.uniforms[uniform];
          if (!u)
            return;
          fsString = insertStringAtIndex(
            fsString,
            `uniform ${u.type} ${uniform}${u.arraySize || ""}; 
`,
            indexOfPrecisionString + precisionString.length
          );
          u.isNeeded = true;
        }
      });
      const isShadertoy = fs.includes("mainImage");
      if (isShadertoy)
        fsString = fsString.concat(FS_MAIN_SHADER);
      return {
        fs: fsString,
        vs
      };
    };
    this.setUniforms = (timestamp) => {
      var _a, _b, _c, _d, _e, _f;
      const { gl } = this;
      if (!gl || !this.shaderProgram)
        return;
      const delta = this.lastTime ? (timestamp - this.lastTime) / 1e3 : 0;
      this.lastTime = timestamp;
      if (this.props.uniforms) {
        Object.keys(this.props.uniforms).forEach((name) => {
          var _a2, _b2;
          const currentUniform = (_a2 = this.props.uniforms) == null ? void 0 : _a2[name];
          if (!currentUniform)
            return;
          if ((_b2 = this.uniforms[name]) == null ? void 0 : _b2.isNeeded) {
            if (!this.shaderProgram)
              return;
            const customUniformLocation = gl.getUniformLocation(
              this.shaderProgram,
              name
            );
            if (!customUniformLocation)
              return;
            processUniform(
              gl,
              customUniformLocation,
              currentUniform.type,
              currentUniform.value
            );
          }
        });
      }
      if ((_a = this.uniforms.iMouse) == null ? void 0 : _a.isNeeded) {
        const mouseUniform = gl.getUniformLocation(
          this.shaderProgram,
          UNIFORM_MOUSE
        );
        gl.uniform4fv(
          mouseUniform,
          this.uniforms.iMouse.value
        );
      }
      if (this.uniforms.iChannelResolution && this.uniforms.iChannelResolution.isNeeded) {
        const channelResUniform = gl.getUniformLocation(
          this.shaderProgram,
          UNIFORM_CHANNELRESOLUTION
        );
        gl.uniform3fv(
          channelResUniform,
          this.uniforms.iChannelResolution.value
        );
      }
      if ((_b = this.uniforms.iDeviceOrientation) == null ? void 0 : _b.isNeeded) {
        const deviceOrientationUniform = gl.getUniformLocation(
          this.shaderProgram,
          UNIFORM_DEVICEORIENTATION
        );
        gl.uniform4fv(
          deviceOrientationUniform,
          this.uniforms.iDeviceOrientation.value
        );
      }
      if ((_c = this.uniforms.iTime) == null ? void 0 : _c.isNeeded) {
        const timeUniform = gl.getUniformLocation(
          this.shaderProgram,
          UNIFORM_TIME
        );
        gl.uniform1f(timeUniform, this.timer += delta);
      }
      if ((_d = this.uniforms.iTimeDelta) == null ? void 0 : _d.isNeeded) {
        const timeDeltaUniform = gl.getUniformLocation(
          this.shaderProgram,
          UNIFORM_TIMEDELTA
        );
        gl.uniform1f(timeDeltaUniform, delta);
      }
      if ((_e = this.uniforms.iDate) == null ? void 0 : _e.isNeeded) {
        const d = /* @__PURE__ */ new Date();
        const month = d.getMonth() + 1;
        const day = d.getDate();
        const year = d.getFullYear();
        const time = d.getHours() * 60 * 60 + d.getMinutes() * 60 + d.getSeconds() + d.getMilliseconds() * 1e-3;
        const dateUniform = gl.getUniformLocation(
          this.shaderProgram,
          UNIFORM_DATE
        );
        gl.uniform4fv(dateUniform, [year, month, day, time]);
      }
      if ((_f = this.uniforms.iFrame) == null ? void 0 : _f.isNeeded) {
        const timeDeltaUniform = gl.getUniformLocation(
          this.shaderProgram,
          UNIFORM_FRAME
        );
        gl.uniform1i(
          timeDeltaUniform,
          this.uniforms.iFrame.value++
        );
      }
      if (this.texturesArr.length > 0) {
        this.texturesArr.forEach((texture, id) => {
          var _a2;
          const { isVideo, _webglTexture, source, flipY, isLoaded } = texture;
          if (!isLoaded)
            return;
          if ((_a2 = this.uniforms[`iChannel${id}`]) == null ? void 0 : _a2.isNeeded) {
            if (!this.shaderProgram)
              return;
            const iChannel = gl.getUniformLocation(
              this.shaderProgram,
              `iChannel${id}`
            );
            gl.activeTexture(gl[`TEXTURE${id}`]);
            gl.bindTexture(gl.TEXTURE_2D, _webglTexture);
            gl.uniform1i(iChannel, id);
            if (isVideo) {
              texture.updateTexture(_webglTexture, source, flipY);
            }
          }
        });
      }
    };
    this.registerCanvas = (r) => {
      this.canvas = r;
    };
    this.mousedown = false;
    this.timer = 0;
    this.lastMouseArr = [0, 0];
    this.texturesArr = [];
    this.lastTime = 0;
    this.render = () => {
      const { style } = this.props;
      const currentStyle = {
        glCanvas: __spreadValues({
          height: "100%",
          width: "100%"
        }, style)
      };
      return (
        // @ts-expect-error TODO: Deal with this.
        /* @__PURE__ */ React.createElement("canvas", { ref: this.registerCanvas, style: currentStyle.glCanvas })
      );
    };
    this.uniforms = {
      [UNIFORM_TIME]: {
        type: "float",
        isNeeded: false,
        value: 0
      },
      [UNIFORM_TIMEDELTA]: {
        type: "float",
        isNeeded: false,
        value: 0
      },
      [UNIFORM_DATE]: {
        type: "vec4",
        isNeeded: false,
        value: [0, 0, 0, 0]
      },
      [UNIFORM_MOUSE]: {
        type: "vec4",
        isNeeded: false,
        value: [0, 0, 0, 0]
      },
      [UNIFORM_RESOLUTION]: {
        type: "vec2",
        isNeeded: false,
        value: [0, 0]
      },
      [UNIFORM_FRAME]: {
        type: "int",
        isNeeded: false,
        value: 0
      },
      [UNIFORM_DEVICEORIENTATION]: {
        type: "vec4",
        isNeeded: false,
        value: [0, 0, 0, 0]
      }
    };
  }
  componentWillUnmount() {
    var _a, _b, _c;
    const { gl } = this;
    if (gl) {
      (_a = gl.getExtension("WEBGL_lose_context")) == null ? void 0 : _a.loseContext();
      gl.useProgram(null);
      gl.deleteProgram((_b = this.shaderProgram) != null ? _b : null);
      if (this.texturesArr.length > 0) {
        this.texturesArr.forEach((texture) => {
          gl.deleteTexture(texture._webglTexture);
        });
      }
      this.shaderProgram = null;
    }
    this.removeEventListeners();
    cancelAnimationFrame((_c = this.animFrameId) != null ? _c : 0);
  }
};
Shader.defaultProps = {
  textures: [],
  contextAttributes: {},
  devicePixelRatio: 1,
  vs: BASIC_VS,
  precision: "highp",
  onError: console.error,
  onWarn: console.warn
};
export {
  ClampToEdgeWrapping,
  LinearFilter,
  LinearMipMapLinearFilter,
  LinearMipMapNearestFilter,
  MirroredRepeatWrapping,
  NearestFilter,
  NearestMipMapLinearFilter,
  NearestMipMapNearestFilter,
  RepeatWrapping,
  Shader
};
